javalib/src/main/scala/java/lang/Long.scala

```
@@ -200,51 +200,48 @@ object Long {
 
+  private def fail(s: String): Nothing =
+    throw new NumberFormatException(s"""For input string: "$s"""")
+
   def decode(nm: String): Long = {
+    if (nm == null)
+      throw new NumberFormatException("null")
     val length = nm.length()
-    if (length == 0) {
-      throw new NumberFormatException()
-    } else {
-      var i          = 0
-      var firstDigit = nm.charAt(i)
-      val negative   = firstDigit == '-'
+    if (length == 0) fail(nm)
 
-      if (negative) {
-        if (length == 1) {
-          throw new NumberFormatException(nm)
-        }
+    var i          = 0
+    var firstDigit = nm.charAt(i)
+    val negative   = firstDigit == '-'
+    val positive   = firstDigit == '+'
 
-        i += 1
-        firstDigit = nm.charAt(i)
-      }
+    if (negative || positive) {
+      if (length == 1) fail(nm)
+      i += 1
+      firstDigit = nm.charAt(i)
+    }
 
-      var base = 10
-      if (firstDigit == '0') {
-        i += 1
-        if (i == length) {
-          return valueOf(0L)
-        }
+    var base = 10
+    if (firstDigit == '0') {
+      i += 1
+      if (i == length) {
+        return valueOf(0L)
+      }
 
-        firstDigit = nm.charAt(i)
-        if (firstDigit == 'x' || firstDigit == 'X') {
-          if (i == length) {
-            throw new NumberFormatException(nm)
-          }
-          i += 1
-          base = 16
-        } else {
-          base = 8
-        }
-      } else if (firstDigit == '#') {
-        if (i == length) {
-          throw new NumberFormatException(nm)
-        }
+      firstDigit = nm.charAt(i)
+      if (firstDigit == 'x' || firstDigit == 'X') {
         i += 1
+        if (i == length) fail(nm)
         base = 16
+      } else {
+        base = 8
       }
-
-      valueOf(parse(nm, i, base, negative))
+    } else if (firstDigit == '#') {
+      i += 1
+      if (i == length) fail(nm)
+      base = 16
     }
+
+    valueOf(parse(nm, i, base, negative))
```

This block has not changed apart from failing with the standard `fail` message that includes the `For input string: "$nm"`. But it has been taken out of an else (and we fail fast instead)